### ISSUE TITLE
Crystal 0.7.6 doesn't compile on Fedora 22 x86_64

### ISSUE THREAD

**Issue by renich** [manastech/crystal#1382]
Hello,

I just tried the following:

```bash
#!/usr/bin/env bash

sudo dnf -y install libxml2-devel gmp-devel libevent-devel pcl-devel pcllib-devel gc-devel libunwind-devel libbsd-devel libyaml-devel readline-devel libedit-devel llvm-static

# settings
path=$( mktemp -d )
os='linux'
arch='x86_64'
version='0.7.6'
release='1'

## tell me where you're working in
echo "working in $path"


# generate tmp dir
cd $path


# get source code
git clone git@github.com:manastech/crystal.git

## change into latest release
cd $path/crystal
git checkout $version
cd $path


# get binary
curl -LO https://github.com/manastech/crystal/releases/download/${version}/crystal-${version}-${release}-${os}-${arch}.tar.gz

## uncompress
tar -xaf crystal-${version}-${release}-${os}-${arch}.tar.gz


# add binary to PATH
export PATH="${path}/crystal-${version}-${release}/bin:${PATH}"


# build
## cleanup
cd $path/crystal
rm -fr .crystal
make clean

## build
make
```

And it ends like this:
```bash
working in /tmp/tmp.EevC2a1vdA
Cloning into 'crystal'...
remote: Counting objects: 63154, done.
remote: Compressing objects: 100% (414/414), done.
remote: Total 63154 (delta 335), reused 0 (delta 0), pack-reused 62721
Receiving objects: 100% (63154/63154), 29.49 MiB | 2.39 MiB/s, done.
Resolving deltas: 100% (42078/42078), done.
Checking connectivity... done.
Note: checking out '0.7.6'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at eb13f75... Crystal 0.7.6
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   413    0   413    0     0    550      0 --:--:-- --:--:-- --:--:--   549
100 10.2M  100 10.2M    0     0  1794k      0  0:00:05  0:00:05 --:--:-- 2411k
rm -rf .build
rm -rf ./doc
CRYSTAL_CONFIG_PATH=`pwd`/src ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr
/tmp/tmp.EevC2a1vdA/crystal/.crystal/tmp/tmp.EevC2a1vdA/crystal/src/ecr/process.cr/main.o: In function `~fun_literal_1':
main_module:(.text+0xf19): undefined reference to `GC_push_all_eager'
main_module:(.text+0xf29): undefined reference to `GC_push_all_eager'
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc -o /tmp/tmp.EevC2a1vdA/crystal/.crystal/crystal-run-macro-run-_tmp_tmp.EevC_a_vdA_crystal_src_ecr_process.cr.tmp "/tmp/tmp.EevC2a1vdA/crystal/.crystal/tmp/tmp.EevC2a1vdA/crystal/src/ecr/process.cr/main.o"   -levent -lrt -lpcl -lpcre -lgc -lpthread -lunwind`
Makefile:25: recipe for target '.build/crystal' failed
make: *** [.build/crystal] Error 3
```

I get the same result in Funtoo; with LLVM v 3.6 and 3.7.

**Comment by jhass** [comment]
You're missing `export LIBRARY_PATH="/opt/crystal/embedded/lib"`

**Comment by renich** [comment]
@jhass ah! Thanks man; sorry. Let me try again.

**Comment by renich** [comment]
OK, still not working. Here's the updated script:

```bash
#!/usr/bin/env bash

# pre-requisites
case $( cat /etc/*release | grep -oiEm 1 'fedora|debian|ubuntu' | awk '{print tolower($0)}' ) in
    fedora)
        echo 'installing prerequisites for Fedora... '

        sudo dnf -y install libxml2-devel gmp-devel libevent-devel pcl-devel pcllib-devel gc-devel libunwind-devel libbsd-devel libyaml-devel readline-devel libedit-devel llvm-static
        ;;
    debian|ubuntu)
        echo 'installing prerequisites for Debian or Ubuntu... '

        sudo apt-get install libbsd-dev libedit-dev libevent-core-2.0-5 libevent-extra-2.0-5 libevent-openssl-2.0-5 libevent-pthreads-2.0-5 libevent-dev libgc-dev libpcl1 libpcl1-dev libunwind8 libunwind8-dev libgmpxx4ldbl libgmp-dev libxml2-dev libyaml-dev readline-dev 
        ;;
    *)
        echo "
I'm sorry. Your OS is not supported by this script. Please, visit:
        
    https://github.com/manastech/crystal/wiki/All-required-libraries

And devise which dependencies you need and expand this script. It is easy enough ;).
"
        exit 1
        ;;
esac

# settings
path=$( mktemp -d )
os='linux'
arch='x86_64'
version='0.7.6'
release='1'

# env
export PATH="${path}/crystal-${version}-${release}/bin:${PATH}"
export LIBRARY_PATH="${path}/crystal-${version}-${release}/embedded/lib"

## tell me where you're working in
echo "working in $path"


# generate tmp dir
cd $path


# get source code
git clone git@github.com:manastech/crystal.git

## change into latest release
cd $path/crystal
git checkout $version
cd $path


# get binary
curl -LO https://github.com/manastech/crystal/releases/download/${version}/crystal-${version}-${release}-${os}-${arch}.tar.gz

## uncompress
tar -xaf crystal-${version}-${release}-${os}-${arch}.tar.gz


# build
## cleanup
cd $path/crystal
rm -fr .crystal
make clean

## build
make
```

and it's output

```bash
installing prerequisites for Fedora... 
[sudo] password for renich: 
Last metadata expiration check performed 2:04:35 ago on Sat Sep  5 13:49:46 2015.
Package libxml2-devel-2.9.2-3.fc22.x86_64 is already installed, skipping.
Package gmp-devel-1:6.0.0-9.fc22.x86_64 is already installed, skipping.
Package libevent-devel-2.0.21-6.fc22.x86_64 is already installed, skipping.
Package pcl-devel-1.7.2-3.fc22.x86_64 is already installed, skipping.
Package pcllib-devel-1.12-3.fc22.x86_64 is already installed, skipping.
Package gc-devel-7.4.2-2.fc22.x86_64 is already installed, skipping.
Package libunwind-devel-1.1-10.fc22.x86_64 is already installed, skipping.
Package libbsd-devel-0.7.0-2.fc22.x86_64 is already installed, skipping.
Package libyaml-devel-0.1.6-6.fc22.x86_64 is already installed, skipping.
Package readline-devel-6.3-5.fc22.x86_64 is already installed, skipping.
Package libedit-devel-3.1-12.20150325cvs.fc22.x86_64 is already installed, skipping.
Package llvm-static-3.5.0-9.fc22.x86_64 is already installed, skipping.
Dependencies resolved.
Nothing to do.
Complete!
working in /tmp/tmp.FtbIqYzl6r
Cloning into 'crystal'...
remote: Counting objects: 63154, done.
remote: Compressing objects: 100% (414/414), done.
remote: Total 63154 (delta 335), reused 0 (delta 0), pack-reused 62721
Receiving objects: 100% (63154/63154), 29.49 MiB | 2.39 MiB/s, done.
Resolving deltas: 100% (42078/42078), done.
Checking connectivity... done.
Note: checking out '0.7.6'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at eb13f75... Crystal 0.7.6
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   411    0   411    0     0    601      0 --:--:-- --:--:-- --:--:--   602
100 10.2M  100 10.2M    0     0  1805k      0  0:00:05  0:00:05 --:--:-- 2151k
rm -rf .build
rm -rf ./doc
CRYSTAL_CONFIG_PATH=`pwd`/src ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr
/tmp/tmp.FtbIqYzl6r/crystal/.crystal/tmp/tmp.FtbIqYzl6r/crystal/src/ecr/process.cr/main.o: In function `~fun_literal_1':
main_module:(.text+0xf19): undefined reference to `GC_push_all_eager'
main_module:(.text+0xf29): undefined reference to `GC_push_all_eager'
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc -o /tmp/tmp.FtbIqYzl6r/crystal/.crystal/crystal-run-macro-run-_tmp_tmp.FtbIqYzl_r_crystal_src_ecr_process.cr.tmp "/tmp/tmp.FtbIqYzl6r/crystal/.crystal/tmp/tmp.FtbIqYzl6r/crystal/src/ecr/process.cr/main.o"   -levent -lrt -lpcl -lpcre -lgc -lpthread -lunwind`
Makefile:25: recipe for target '.build/crystal' failed
make: *** [.build/crystal] Error 3
```

**Comment by jhass** [comment]
Okay, looks like it prefers the globally installed libgc.so over the libgc.a we ship in the package unless you explicitly tell it to use the static version. So until https://github.com/ivmai/bdwgc/commit/1a49011e0ab0ab29df237eec8556df29c9af9c31 is in a widespread release, you may have to workaround these with a patch like

```diff
--- a/src/gc/boehm.cr
+++ b/src/gc/boehm.cr
@@ -1,5 +1,5 @@
 @[Link("pthread")]
-@[Link("gc")]
+@[Link(ldflags: "-Wl,-Bstatic -lgc -Wl,-Bdynamic")]
 lib LibGC
   fun init = GC_init
   fun malloc = GC_malloc(size : UInt32) : Void*
```

or maybe somebody has something clever we can pass via `CFLAGS` or `--link-flags`.

**Comment by renich** [comment]
OK, I think it didn't work. Here's my proof:

```bash
[renich@introlap crystal]$ pwd
/tmp/tmp.4nHmMdkslI/crystal
[renich@introlap crystal]$ head ../crystal-0.7.7-1/src/gc/boehm.cr
@[Link("pthread")]
@[Link(ldflags: "-Wl,-Bstatic -lgc -Wl,-Bdynamic")]
lib LibGC
  alias Int = LibC::Int
  alias SizeT = LibC::SizeT
  alias Word = LibC::ULong

  fun init = GC_init
  fun malloc = GC_malloc(size : SizeT) : Void*
  fun malloc_atomic = GC_malloc_atomic(size : SizeT) : Void*
[renich@introlap crystal]$ echo $LIBRARY_PATH 
/tmp/tmp.4nHmMdkslI/crystal-0.7.7-1/embedded/lib
[renich@introlap crystal]$ echo $PATH 
/tmp/tmp.4nHmMdkslI/crystal-0.7.7-1/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/home/renich/.local/bin:/home/renich/bin
[renich@introlap crystal]$ make
CRYSTAL_CONFIG_PATH=`pwd`/src ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr
./.crystal/tmp/tmp.4nHmMdkslI/crystal/src/ecr/process.cr/main.o: In function `~fun_literal_1':
main_module:(.text+0xfb9): undefined reference to `GC_push_all_eager'
main_module:(.text+0xfc9): undefined reference to `GC_push_all_eager'
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc -o /tmp/tmp.4nHmMdkslI/crystal/.crystal/crystal-run-macro-run-_tmp_tmp._nHmMdkslI_crystal_src_ecr_process.cr.tmp "./.crystal/tmp/tmp.4nHmMdkslI/crystal/src/ecr/process.cr/main.o"   -levent -lrt -lpcl -lpcre -lgc -lpthread -lunwind`
Makefile:25: recipe for target '.build/crystal' failed
make: *** [.build/crystal] Error 1
```

Patch was applied and it didn't work anyway. Let me know if you have any other ideas.

**Comment by jhass** [comment]
You must patch the source you compile, not the one shipped with the download.

**Comment by lbguilherme** [comment]
I have run into the same problem. The fix that worked for me was fetching libgc from source (master already have the patch, although there is no release of it yet), building it and installing it system-wide.

```
git clone git://github.com/ivmai/bdwgc.git
cd bdwgc
git clone git://github.com/ivmai/libatomic_ops.git
autoreconf -vif
automake --add-missing
./configure
make
make check
sudo make install
```

Maybe it is not very clear, as it will include other non-released changes that may break something, but it has served me well.

**Comment by jhass** [comment]
This is actually related to #1269

**Comment by renich** [comment]
ah, great! OK, I will wait for the fix.

**Comment by jhass** [comment]
Well, I'm not sure if anybody is actively working on it, having the workaround would certainly be preferable.

**Comment by renich** [comment]
@jhass well, could we tag it as a bug?

IMHO, overall distro adoption is of paramount importance. I think this issue should be taken more seriously.

I know you distribute binaries. That is never the same as passing through a distribution's scrutiny and revision.

**Comment by jhass** [comment]
We will probably release 0.7.8 today or tomorrow, alongside we will ship `-fPIC`  compatible libraries.

**Comment by renich** [comment]
That sounds really good. Will standby and confirm the successful build.

**Comment by jhass** [comment]
But forcing a static link for bdw-gc will break for example the Archlinux package, so it's either you applying the patch or Arch applying the patch (and probably more people given there have been rather few complaints about this issue so far).

**Comment by renich** [comment]
I'll apply the patch, then. Fedora isn't big on static linking anyway. I dunno what Gentoo/Funtoo's opinion on this is, though.

**Comment by renich** [comment]
Just as a reference on static linking: 

Fedora: https://fedoraproject.org/wiki/Packaging:Guidelines#Packaging_Static_Libraries_2
OpenSuse: https://en.opensuse.org/openSUSE:Packaging_guidelines#Static_Libraries
Gentoo: https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies
Debian: https://wiki.debian.org/StaticLinking

As I understand it, they all advise against static linking.

**Comment by jhass** [comment]
To summarize it:

* We depend on a function of bdw-gc that is not exposed in the current release.
* We ship an archive file of bdw-gc for static linking.
* When compiling said archive file, the function is exposed anyway.
* When only compiling the dynamic library of bdw-gc, like most distros do, the function is not exposed.
* bdw-gc made a patch for us that exposes the function, but that patch is not in a release yet.
* Archlinux applies that patch in it's package for bdw-gc but does not enable building the archive file.
* When there's a shared library to be found by the linker it's preferred over the archive file unless static linking is explicitly requested.

So you have three options:

* Force static linking of bdw-gc (that's what my patch above does) and make sure `LIBRARY_PATH` contains the directory containing the archive file we ship.
* Get the patch that exposes the necessary functions into the distro packages for bdw-gc you care about.
* Convince the the distros you care about to rebuild their bdw-gc packages without disabling the build of the archive file, this exposes the necessary functions in the shared library too, as an unintended side effect.

All of these would be a temporary measure until the next bdw-gc release happened and is widespread, however 2, 3 and are significantly harder, since you need to convince some distro maintainers.

**Comment by renich** [comment]
I appreciate the summary a lot. Thanks.

**Comment by renich** [comment]
Hello, still testing on Fedora 22 but now with version 0.8.0. It compiles with the patch but it doesn't run:

```bash
!#/usr/bin/env bash

# settings
path=$( mktemp -d )
scriptpath=$( readlink -f $( dirname $0 ) )
os='linux'
arch='x86_64'
version='0.8.0'
release='1'

# env
export PATH="${path}/crystal-${version}-${release}/bin:${PATH}"
export LIBRARY_PATH="${path}/crystal-${version}-${release}/embedded/lib"

## tell me where you're working in
echo "working in $path"


# generate tmp dir
cd $path


# get source code
git clone git@github.com:manastech/crystal.git

## change into latest release
cd $path/crystal
git checkout $version

## patch
patch -p1 < $scriptpath/bdw-gc-temporary_static_link.patch || exit 1

cd $path


# get binary
curl -LO https://github.com/manastech/crystal/releases/download/${version}/crystal-${version}-${release}-${os}-${arch}.tar.gz

## uncompress
tar -xaf crystal-${version}-${release}-${os}-${arch}.tar.gz


# build
## cleanup
cd $path/crystal
rm -fr .crystal
make clean

## build
make

## test
cat << EOF > $path/crystal/test.cr
puts "Hello, world!"
EOF

crystal run $path/crystal/test.cr
```

And this is the output

```bash
[renich@introlap crystal]$ ./do 
installing prerequisites for Fedora... 
[sudo] password for renich: 
Sorry, try again.
[sudo] password for renich: 
Last metadata expiration check performed 0:38:57 ago on Tue Sep 22 19:14:31 2015.
Package libxml2-devel-2.9.2-4.fc22.x86_64 is already installed, skipping.
Package gmp-devel-1:6.0.0-9.fc22.x86_64 is already installed, skipping.
Package libevent-devel-2.0.21-6.fc22.x86_64 is already installed, skipping.
Package pcl-devel-1.7.2-3.fc22.x86_64 is already installed, skipping.
Package pcllib-devel-1.12-3.fc22.x86_64 is already installed, skipping.
Package gc-devel-7.4.2-2.fc22.x86_64 is already installed, skipping.
Package libunwind-devel-1.1-10.fc22.x86_64 is already installed, skipping.
Package libbsd-devel-0.7.0-2.fc22.x86_64 is already installed, skipping.
Package libyaml-devel-0.1.6-6.fc22.x86_64 is already installed, skipping.
Package readline-devel-6.3-5.fc22.x86_64 is already installed, skipping.
Package libedit-devel-3.1-12.20150325cvs.fc22.x86_64 is already installed, skipping.
Package llvm-static-3.5.0-9.fc22.x86_64 is already installed, skipping.
Dependencies resolved.
Nothing to do.
Complete!
working in /tmp/tmp.1GDkiJApjt
Cloning into 'crystal'...
remote: Counting objects: 66964, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 66964 (delta 0), reused 0 (delta 0), pack-reused 66961
Receiving objects: 100% (66964/66964), 32.04 MiB | 388.00 KiB/s, done.
Resolving deltas: 100% (45095/45095), done.
Checking connectivity... done.
Note: checking out '0.8.0'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at e363b63... Crystal 0.8.0
patching file src/gc/boehm.cr
Hunk #1 succeeded at 1 with fuzz 2.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   596    0   596    0     0    690      0 --:--:-- --:--:-- --:--:--   690
100 10.4M  100 10.4M    0     0   305k      0  0:00:35  0:00:35 --:--:-- 1509k
rm -rf .build
rm -rf ./doc
rm -rf src/llvm/ext/llvm_ext.o
g++ -c -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc `/usr/bin/llvm-config --cxxflags`
CRYSTAL_CONFIG_PATH=`pwd`/src ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr
./.crystal/tmp/tmp.1GDkiJApjt/crystal/test.cr/Fiber.o: In function `*Fiber#push_gc_roots<Fiber>:Void':
Fiber:(.text+0x17): undefined reference to `GC_push_all_eager'
Fiber:(.text+0x4c): undefined reference to `GC_push_all_eager'
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc -o "/tmp/tmp.1GDkiJApjt/crystal/.crystal/crystal-run-test.tmp" "${@}"  -rdynamic  -levent -lrt -lpcl -lpcre -lgc -lpthread -ldl`
```

**Comment by jhass** [comment]
`crystal` there resolves to the tarball and thus its stdlib. You want `bin/crystal` to get the one you compiled (and its stdlib).

**Comment by jhass** [comment]
No further feedback for almost 4 weeks, closing.
